### DISCUSSION TITLE
understanding-default-interface-methods/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#407] [Announcements]
# understanding-default-interface-methods/

In this post I provide an introduction to default interface methods, how they work at a high level, their typical uses, and some of their sharp edges

https://andrewlock.net/understanding-default-interface-methods/

**Comment by Balinski** [comment]
Thanks, perfect and easily absorbed material feed! Thank again!

**Comment by Mrxx99** [comment]
Nice article
Small mistake in 6th code snippet: it does not compile with the override keyword

**Comment by AndreiGorlov** [comment]
The code in "The diamond inheritance problem" compiles fine with warnings:

'IHasCurvedEdges.GetName()' hides inherited member 'IShape.GetName()'. Use the new keyword if hiding was intended.
'IHasStraightEdges.GetName()' hides inherited member 'IShape.GetName()'. Use the new keyword if hiding was intended.

Correct example:

``` csharp
interface IShape
{
    string GetName() => "IShape"; // Default implementation
}

interface IHasStraightEdges : IShape
{
    string IShape.GetName() => "IHasStraightEdges"; // Override the default
}

interface IHasCurvedEdges : IShape
{
    string IShape.GetName() => "IHasCurvedEdges"; // Override the default
}

// inherits both interfaces
public class MySemiCircle : IHasStraightEdges, IHasCurvedEdges {} // error
```

and fix is:
``` csharp
public class MySemiCircle : IHasStraightEdges, IHasCurvedEdges
{
    public string GetName() => "MySemiCircle"; // provide an implementation
}
```
without inherit from IShape explicitly

**Reply by andrewlock** [reply]
Yeah, you're right, not sure what happened there 😅 Fix going out now, thanks

**Comment by farkaszsombor** [comment]
Thank you for another insightful post!

Aren't "extension methods" are an better alternative for extending an interface ? I feel this default interface thing a little too weird.

**Comment by taconaut** [comment]
Maybe it will just take some time for adjustment, but default implementations change fundamentally how we must look at interfaces. Up to now, it was a strict contract, e.g. the IDE would create methods for when implementing one. Now we might have 'hidden' features to pay attention to. When a feature gets added, one might miss it; is it important? Can/should it remain default?
Not convinced to like this more blurry option over the previous clarity.
Not writing public libs though.

**Reply by ayodejii** [reply]
> Now we might have 'hidden' features to pay attention to

this is true, but i think we don't necessarily need to pay attention to these features if we do not need them.

This guy on reddit kinda looks at your concern: https://www.reddit.com/r/csharp/comments/120kr4p/comment/jdi24pe/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Comment by ayodejii** [comment]
explanation and example too good!
Thank you!

**Comment by KhorAMus** [comment]
EF core has interface called ISaveChangesInterceptor. It has default implementations for all methods. This methods are like event handlers. And you can implement only some of them. So you can use default implementations in this case too.

**Comment by MauNguyenVan** [comment]
Take a look at this code:
```cs
internal class DefaultInterfaceMemberBug
{
  public static void Main()
  {
    M(new C1()); // Print: 12 
    M(new C2()); // Print: ??, Can you guess?
  }

  static void M(I x)
  {
    x.M1();
    x.M2();
  }
}

partial interface I
{
  public void M1()
  {
    Console.Write(1);
  }

  public partial void M2();
  public partial void M2()
  {
    Console.Write(2);
  }
}

class C1 : I;

class C2 : I
{
  public void M1()
  {
    Console.Write(3);
  }

  public void M2()
  {
    Console.Write(4);
  }
}
```
Now, logically you might expect the output to be:
12  
34
But that's not what happens. Instead, the output is:
12  
32
This is not just limited to methods—it can happen with partial properties and partial indexers too.

This issue is being tracked here:
🔗 github.com/dotnet/roslyn/issues/77346

If you’re working with C# interfaces and using newer features like partial or default members, definitely check this out. It’s a subtle bug that could trip you up. 

. Net team will fix for partial property and partial even, but not for partial method, this fix will make inconsitancy behavior
if you want to Design team fix all of it, please go to github link to comment and vote for fixing method

**Comment by Ungerfall** [comment]
Do you know why "implement interface" in visual studio does not include default methods?
This is very confusing and caused a bug for me:
I was writing a `Scruptor` decorator for `IMemoryCache`. The visual studio didn't add `GetCurrentStatistics`, which caused `GetCurrentStatistics` to return `null` even with `services.AddMemoryCache(setup => setup.TrackStatistics = true)`
